src/extensions/scratch3_video_sensing/index.js

```diff
@@ -0,0 +1,336 @@
+
+/**
+ * Class for the motion-related blocks in Scratch 3.0
+ * @param {Runtime} runtime - the runtime instantiating this block package.
+ * @constructor
+ */
+class Scratch3VideoSensingBlocks {
+    constructor (runtime) {
+        /**
+         * The runtime instantiating this block package.
+         * @type {Runtime}
+         */
+        this.runtime = runtime;
+
+        this.detect = new VideoMotion();
+
+        this._lastUpdate = null;
+
+        this._skinId = -1;
+        this._skin = null;
+        this._drawable = -1;
+
+        this._setupVideo();
+        this._setupSampleCanvas();
+        this._setupPreview();
+        this._loop();
+    }
+
+    static get INTERVAL () {
+        return 33;
+    }
+
+    static get DIMENSIONS () {
+        return [480, 360];
+    }
+
+    static get ORDER () {
+        return 1;
+    }
+
+    _setupVideo () {
+        this._video = document.createElement('video');
+        navigator.getUserMedia({
+            audio: false,
+            video: {
+                width: {min: 480, ideal: 640},
+                height: {min: 360, ideal: 480}
+            }
+        }, stream => {
+            this._video.src = window.URL.createObjectURL(stream);
+            // Hint to the stream that it should load. A standard way to do this
+            // is add the video tag to the DOM. Since this extension wants to
+            // hide the video tag and instead render a sample of the stream into
+            // the webgl rendered Scratch canvas, another hint like this one is
+            // needed.
+            this._track = stream.getTracks()[0];
+        }, err => {
+            // @todo Properly handle errors
+            log(err);
+        });
+    }
+
+    _setupSampleCanvas () {
+        // Create low-resolution image to sample video for analysis and preview
+        const canvas = this._sampleCanvas = document.createElement('canvas');
+        canvas.width = Scratch3VideoSensingBlocks.DIMENSIONS[0];
+        canvas.height = Scratch3VideoSensingBlocks.DIMENSIONS[1];
+        this._sampleContext = canvas.getContext('2d');
+    }
+
+    _setupPreview () {
+        if (this._skinId !== -1) return;
+        if (this._skin !== null) return;
+        if (this._drawable !== -1) return;
+        if (!this.runtime.renderer) return;
+
+        this._skinId = this.runtime.renderer.createPenSkin();
+        this._skin = this.runtime.renderer._allSkins[this._skinId];
+        this._drawable = this.runtime.renderer.createDrawable();
+        this.runtime.renderer.setDrawableOrder(
+            this._drawable,
+            Scratch3VideoSensingBlocks.ORDER
+        );
+        this.runtime.renderer.updateDrawableProperties(this._drawable, {
+            skinId: this._skinId
+        });
+    }
+
+    _loop () {
+        setTimeout(this._loop.bind(this), this.runtime.THREAD_STEP_INTERVAL_COMPATIBILITY);
```

Comment:
Should this use `Scratch3VideoSensingBlocks.INTERVAL` instead? Or perhaps it should be `runtime.currentStepTime`?

Reply:
👍 That is actually my bad from the Vision extension. Sorry @mzgoddard !

Reply:
That's a good point @cwillisf. Eventually we probably want to use [`timeupdate`](https://developer.mozilla.org/en-US/docs/Web/Events/timeupdate) or some thing to execute a step of the loop approximately at the hertz of the camera.